Pathfinding cleanup

ZehMatt opened the pull request:
Refactored the strange abstraction, eliminated a few globals/statics and applied our code style, each individual commit was tested against our replays so there shouldn't be anything wrong. I personally tested some cases like staff ignoring banners and making sure guests don't, I don't think replays cover this properly.

I plan to do more work in this area while trying to keep PRs as small as possible.

ZehMatt commented:
> > Eliminate the ridiculous abstraction for path finding
> 
> I think this is an unprofessional commit message to begin with. But also: the abstraction is there so that the algorithm can be easily exchanged for a better one. It’s hardly ‘strange’ or ‘ridiculous’.

This is abstraction done wrong, nearly 70% of the code has nothing to do with actual path finding. Abstraction only makes sense when there is a clear separation of concerns. It is also questionable to have more than one path finding system and abstraction also comes with a cost. I may not have chosen the most political correct wording and I can change it but my opinion about it remains the same.

Gymnasiast commented:
> It is also questionable to have more than one path finding system and abstraction also comes with a cost.

It is hardly questionable. OpenTTD used the same pattern it to trial their new pathfinder for years before being confident enough to drop the old one. This was also the reason for abstracting it here - being able to develop a new, better pathfinder and being able to trial and test it while keeping the existing code working.

I cannot vouch for the implementation - you know much about that than I do - but the idea is hardly strange. And the need for being able to get a better pathfinding system in the future and being able to test it hasn’t gone away, so it would have to be addressed in _some_ way if the current implementation is not up to snuff.

LordOfLunacy commented:
Hi, I was looking at the codebase because I was curious if anything was done with this abstraction, after I was pulled away from doing anything with it. Admittedly, this was my first time interacting with this codebase so I wouldn't be surprised if the way I implemented the pathfinding object was not ideal, however skimming through these changes besides the removal of globals (which was something I had a ticket for that also introduced an additional replay test iirc that wasnt merged before I got pulled away), I do not really understand what is better about the changes you made, as it mostly reverts the code back to how I found it without really changing where I see things that could have been done better maybe, but please correct me if I'm missing something, as I am always open to critique that helps me become a better programmer.

The actual pathfinding object template I personally don't see how any of that is unrelated to the actual pathfinding itself, it exposes the things that any pathfinding implementation would need. However, the "OriginalPathfinding" class is something I could see being done better, and contains things that aren't always things that should be implemented as part of the pathfinding itself, but instead should be a function call to another part of the codebase perhaps. However, that is due to how the code was written originally moreso than what I did. I woukd say a refactoring of this part of the code would be in order.

ZehMatt commented:
> Hi, I was looking at the codebase because I was curious if anything was done with this abstraction, after I was pulled away from doing anything with it. Admittedly, this was my first time interacting with this codebase so I wouldn't be surprised if the way I implemented the pathfinding object was not ideal, however skimming through these changes besides the removal of globals (which was something I had a ticket for that also introduced an additional replay test iirc that wasnt merged before I got pulled away), I do not really understand what is better about the changes you made, as it mostly reverts the code back to how I found it without really changing where I see things that could have been done better maybe, but please correct me if I'm missing something, as I am always open to critique that helps me become a better programmer.
> 
> The actual pathfinding object template I personally don't see how any of that is unrelated to the actual pathfinding itself, it exposes the things that any pathfinding implementation would need. However, the "OriginalPathfinding" class is something I could see being done better, and contains things that aren't always things that should be implemented as part of the pathfinding itself, but instead should be a function call to another part of the codebase perhaps. However, that is due to how the code was written originally moreso than what I did. I woukd say a refactoring of this part of the code would be in order.

I want to first focus on actually refactoring the existing code before trying to make any drastic changes or even attempt to have some form of abstraction of this, the virtual calls do absolutely nothing in its current state except making the compiler emit worse code, so there is currently no benefit whatsoever to having this. I don't mind people throwing all sort of funky algorithms in there once its actually done in a way where the algorithm can be easily replaced.

LordOfLunacy commented:
I think that's fair potentially. I am unsure whether or not the compiler is actually spitting out significantly worse code due to the abstraction, the functions are marked final override which should help the compiler get some of that back. Would be curious to see if you benchmarked this, but I agree these functions are called extremely often and in busier parks can take a good bit of CPU time, so if there's significant truth to this slowing things down then that alone may be worth going a different route here. 

But besides that, I viewed this more as a chicken and egg problem, what comes first; a major rework of the pathfinding, or an interface that allows it to be developed and tested independently of the current. The current API was intended to be minimal amd expose only the bare minimum that needed to be so to leave lots of room for growth and changes internal to the implementation.

frutiemax92 commented:
You can use some templates so you keep the benefit of correctly isolating the pathfinding algorithm without using abstraction. I agree that abstraction shouldn't be used for a non-polymorphic/static use as this screws up the cache.

ZehMatt commented:
> I think that's fair potentially. I am unsure whether or not the compiler is actually spitting out significantly worse code due to the abstraction, the functions are marked final override which should help the compiler get some of that back. Would be curious to see if you benchmarked this, but I agree these functions are called extremely often and in busier parks can take a good bit of CPU time, so if there's significant truth to this slowing things down then that alone may be worth going a different route here.
> 
> But besides that, I viewed this more as a chicken and egg problem, what comes first; a major rework of the pathfinding, or an interface that allows it to be developed and tested independently of the current. The current API was intended to be minimal amd expose only the bare minimum that needed to be so to leave lots of room for growth and changes internal to the implementation.

I understand the reasoning behind it however I have to disagree with how its done, we can not and should not built an abstraction based on an already convoluted base. I will continue to work on this and once that is done we can think about how to potentially provide another model next to it, at the end of the day it should be a simple algorithm that provides the information on how to get from A to B, may that be A* or something else, the current code is too deeply connected with some of the peep state.

LordOfLunacy commented:
I will trust your judgment here. I agree that there's some significant reworking of the original pathfinding that needs to be done, and that it's doing a lot of things it shouldn't be doing that goes beyond just pathfinding, and there's certainly more than one correct way to go about it, so the one who's making the changes should be the one who decides that.

Going to say something I'm sure you already know and have thought about, but make sure that when you're making these changes that you don't inadvertently close off potential for some of the likely improvements that would want to be made, such as guests using transportation rides for, you know, transport (not that the code is well suited to this expansion in its current state either). 

I do think long-term the project is going to want the option to choose between the current/original pathfinding and an improved version that can see more than the current 16 or so waypoints it can do currently before falling back to compass navigation. But maybe it is best that waits for guest behavior to be better separated out.

Gymnasiast commented on src/openrct2/peep/GuestPathfinding.cpp:
```diff
@@ -1992,8 +1990,7 @@ namespace OpenRCT2::PathFinding
 
-        _peepPathFindIsStaff = false;
-        uint8_t edges = PathGetPermittedEdges(pathElement);
+        uint8_t edges = PathGetPermittedEdges(false, pathElement);
```
See previous.

ZehMatt replied:
See comment above.

Gymnasiast commented on src/openrct2/peep/GuestPathfinding.cpp:
```diff
@@ -688,7 +688,7 @@ namespace OpenRCT2::PathFinding
     static void PeepPathfindHeuristicSearch(
         TileCoordsXYZ loc, const TileCoordsXYZ& goal, const Peep& peep, TileElement* currentTileElement,
-        const bool inPatrolArea, uint8_t counter, uint16_t* endScore, Direction test_edge, uint8_t* endJunctions,
+        const bool inPatrolArea, uint8_t steps, uint16_t* endScore, Direction testEdge, uint8_t* endJunctions,
```
I’d prefer to rename `steps` to `numSteps`.

ZehMatt commented on src/openrct2/peep/GuestPathfinding.cpp:
```diff
@@ -453,7 +452,7 @@ namespace OpenRCT2::PathFinding
         }
 
-        return FootpathElementDestInDir(loc, chosenDirection, outRideIndex, 0);
+        return FootpathElementDestInDir(false, loc, chosenDirection, outRideIndex, 0);
```
This function is only called for guests, I will add a comment.
